src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      Object.assign(currentState, action.extraData);
+    }
+
+    if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+
+      for (const key of action.keysToRemove) {
+        if (currentState.hasOwnProperty(key)) {
+          delete currentState[key];
+        }
+      }
+    }
+
+    if (action.type === 'clear') {
+      currentState = {};
+    }
+
+    states.push({ ...currentState });
```

Comment:
Consider using a `switch` statement instead of multiple `if` statements for better readability and to adhere to the checklist requirement for a limited amount of conditions.